// @atlaskit/field-base/dist/esm/components/ValidationElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

/* eslint-disable react/prop-types */
import styled from 'styled-components';
import React, { Component } from 'react';
import { yellow } from '@atlaskit/theme/colors';
import Spinner from '@atlaskit/spinner';
import Icon from '@atlaskit/icon/glyph/warning'; // exported for testing

export var WarningIcon = styled.div.withConfig({
  displayName: "ValidationElement__WarningIcon",
  componentId: "sc-1fk9yms-0"
})(["\n  align-items: center;\n  color: ", ";\n  display: flex;\n  flex-shrink: 0;\n"], yellow); // Spinner needs set height to avoid height jumping
// Also needs a margin so there is space between it and preceding text

var SpinnerParent = styled.div.withConfig({
  displayName: "ValidationElement__SpinnerParent",
  componentId: "sc-1fk9yms-1"
})(["\n  height: 20px;\n  margin-left: 10px;\n"]);

var ValidationElement = /*#__PURE__*/function (_Component) {
  _inherits(ValidationElement, _Component);

  var _super = _createSuper(ValidationElement);

  function ValidationElement() {
    _classCallCheck(this, ValidationElement);

    return _super.apply(this, arguments);
  }

  _createClass(ValidationElement, [{
    key: "render",
    value: function render() {
      if (!this.props.isDisabled && this.props.isInvalid) {
        return /*#__PURE__*/React.createElement(WarningIcon, null, /*#__PURE__*/React.createElement(Icon, {
          label: "warning"
        }));
      }

      return this.props.isLoading ? /*#__PURE__*/React.createElement(SpinnerParent, null, /*#__PURE__*/React.createElement(Spinner, {
        size: "small"
      })) : null;
    }
  }]);

  return ValidationElement;
}(Component);

_defineProperty(ValidationElement, "defaultProps", {
  isDisabled: false,
  isInvalid: false,
  isLoading: false
});

export { ValidationElement as default };
