// @compiled/react/dist/cjs/codemods/styled-components-to-compiled/styled-components-to-compiled.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var codemods_helpers_1 = require("../codemods-helpers");
var imports = {
    compiledStyledImportName: 'styled',
    styledComponentsPackageName: 'styled-components',
};
var transformer = function (fileInfo, _a, options) {
    var j = _a.jscodeshift;
    var source = fileInfo.source;
    var collection = j(source);
    var hasStyledComponentsImportDeclaration = codemods_helpers_1.hasImportDeclaration({
        j: j,
        collection: collection,
        importPath: imports.styledComponentsPackageName,
    });
    if (!hasStyledComponentsImportDeclaration) {
        return source;
    }
    codemods_helpers_1.addCommentForUnresolvedImportSpecifiers({
        j: j,
        collection: collection,
        importPath: imports.styledComponentsPackageName,
        allowedImportSpecifierNames: [],
    });
    codemods_helpers_1.convertDefaultImportToNamedImport({
        j: j,
        collection: collection,
        importPath: imports.styledComponentsPackageName,
        namedImport: imports.compiledStyledImportName,
    });
    return collection.toSource(options.printOptions || { quote: 'single' });
};
exports.default = transformer;
