// @atlaskit/tag/dist/esm/constants.js
import { borderRadius, fontSize, gridSize } from '@atlaskit/theme/constants';
var gridSizeValue = gridSize();
var borderRadiusValue = borderRadius();
var fontSizeValue = fontSize();
export var tagHeightUnitless = 2.5 * gridSizeValue;
export var tagHeight = "".concat(tagHeightUnitless, "px");
export var buttonWidthUnitless = tagHeightUnitless; // button should be square

export var buttonWidth = tagHeight; // button should be square

export var maxWidthUnitless = 25 * gridSizeValue;
export var maxWidth = "".concat(maxWidthUnitless, "px");
export var maxTextWidthUnitless = maxWidthUnitless - tagHeightUnitless;
export var maxTextWidth = "".concat(maxTextWidthUnitless, "px");
export var defaultBorderRadius = "".concat(borderRadiusValue, "px");
export var defaultRoundedBorderRadius = "".concat(buttonWidthUnitless / 2, "px");
export var defaultMargin = "".concat(gridSizeValue / 2, "px");
export var defaultTextPadding = "".concat(gridSizeValue / 2, "px");
export var textPaddingRight = "".concat(2 * gridSizeValue, "px");
export var textMarginLeft = "".concat(tagHeightUnitless, "px");
export var textFontSize = "".concat(fontSizeValue, "px");
export var cssVar = {
  color: {
    background: {
      default: '--ds-cb',
      active: '--ds-cba',
      hover: '--ds-cbh'
    },
    focusRing: '--ds-cfr',
    text: {
      default: '--ds-ct',
      hover: '--ds-cth'
    }
  },
  borderRadius: '--ds-br'
};
