// @atlaskit/checkbox/dist/es2019/performance/examples.js
import React from 'react';
import { fireEvent } from '@testing-library/react';
import { Checkbox } from '../../src';
const interactionTasks = [{
  name: 'Select checkbox',
  description: 'Select checkbox',
  run: async ({
    container,
    controls
  }) => {
    const checkbox = container.querySelector('input[name="checkbox-basic"]');

    if (checkbox == null) {
      throw new Error('Could not find checkbox element');
    }

    await controls.time(async () => {
      fireEvent.click(checkbox);
    });
  }
}, {
  name: 'Hover checkbox',
  description: 'Hover checkbox',
  run: async ({
    container,
    controls
  }) => {
    const checkbox = container.querySelector('input[name="checkbox-basic"]');

    if (checkbox == null) {
      throw new Error('Could not find checkbox element');
    }

    await controls.time(async () => {
      fireEvent.mouseOver(checkbox);
    });
  }
}, {
  name: 'Focus checkbox',
  description: 'Focus checkbox',
  run: async ({
    container,
    controls
  }) => {
    const checkbox = container.querySelector('input[name="checkbox-basic"]');

    if (checkbox == null) {
      throw new Error('Could not find checkbox element');
    }

    await controls.time(async () => {
      fireEvent.focus(checkbox);
    });
  }
}];

function PerformanceComponent() {
  return /*#__PURE__*/React.createElement(Checkbox, {
    value: "Basic checkbox",
    label: "Basic checkbox",
    name: "checkbox-basic",
    testId: "cb-basic"
  });
}

export const performance = () => /*#__PURE__*/React.createElement(PerformanceComponent, null);
performance.story = {
  name: 'Performance',
  parameters: {
    performance: {
      interactions: interactionTasks
    }
  }
};
