// @giphy/react-native-sdk/lib/commonjs/GiphyVideoView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GiphyVideoView = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _GiphyVideoView = require("./native/GiphyVideoView");

var _GiphyVideoManager = require("./GiphyVideoManager");

var _deprecatedPropType = _interopRequireDefault(require("./utils/deprecatedPropType"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

let mountedComponentsCount = 0;
let appStateListenerAdded = false;
let latestAppState = _reactNative.AppState.currentState;

function appStateListener(appState) {
  if (latestAppState === 'active' && appState.match(/inactive|background/)) {
    _GiphyVideoManager.GiphyVideoManager.muteAll();
  }

  latestAppState = appState;
}

function addAppStateListener() {
  latestAppState = _reactNative.AppState.currentState;

  _reactNative.AppState.addEventListener('change', appStateListener);

  appStateListenerAdded = true;
}

function removeAppStateListener() {
  _reactNative.AppState.addEventListener('change', appStateListener);

  appStateListenerAdded = false;
}

function registerComponent() {
  mountedComponentsCount += 1;

  if (!appStateListenerAdded) {
    addAppStateListener();
  }
}

function unregisterComponent() {
  mountedComponentsCount = Math.max(0, mountedComponentsCount - 1);

  if (mountedComponentsCount === 0 && appStateListenerAdded) {
    removeAppStateListener();
  }
}

class GiphyVideoView extends _react.default.Component {
  componentDidMount() {
    registerComponent();
  }

  componentWillUnmount() {
    unregisterComponent();
  }

  render() {
    return /*#__PURE__*/_react.default.createElement(_GiphyVideoView.NativeGiphyVideoView, this.props);
  }

}

exports.GiphyVideoView = GiphyVideoView;

_defineProperty(GiphyVideoView, "propTypes", {
  playing: (0, _deprecatedPropType.default)({
    explanation: 'This property will be removed in v2, please use autoPlay instead'
  })
});
