// @atlaskit/modal-dialog/dist/cjs/components/Footer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireDefault(require("react"));

var _customThemeButton = _interopRequireDefault(require("@atlaskit/button/custom-theme-button"));

var _Content = require("../styled/Content");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var JustifyShim = function JustifyShim(props) {
  return /*#__PURE__*/_react.default.createElement("span", props);
};

var ModalFooter = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(ModalFooter, _React$Component);

  var _super = _createSuper(ModalFooter);

  function ModalFooter() {
    (0, _classCallCheck2.default)(this, ModalFooter);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ModalFooter, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          actions = _this$props.actions,
          appearance = _this$props.appearance,
          component = _this$props.component,
          onClose = _this$props.onClose,
          showKeyline = _this$props.showKeyline;
      var warning = 'You can provide `component` OR `actions`, not both.';

      if (!component && !actions) {
        return null;
      }

      if (component && actions) {
        console.warn(warning); // eslint-disable-line no-console

        return null;
      }

      if (component) {
        return /*#__PURE__*/_react.default.createElement(component, {
          appearance: appearance,
          onClose: onClose,
          showKeyline: showKeyline
        });
      }

      return /*#__PURE__*/_react.default.createElement(_Content.Footer, {
        showKeyline: showKeyline
      }, /*#__PURE__*/_react.default.createElement(JustifyShim, null), /*#__PURE__*/_react.default.createElement(_Content.Actions, null, actions ? actions.map(function (_ref, index) {
        var text = _ref.text,
            rest = (0, _objectWithoutProperties2.default)(_ref, ["text"]);
        var variant = index !== 0 ? 'subtle' : appearance || 'primary';
        return /*#__PURE__*/_react.default.createElement(_Content.ActionItem, {
          key: index
        }, /*#__PURE__*/_react.default.createElement(_customThemeButton.default, (0, _extends2.default)({
          appearance: variant
        }, rest), text));
      }) : null));
    }
  }]);
  return ModalFooter;
}(_react.default.Component);

exports.default = ModalFooter;
