// @atlaskit/modal-dialog/dist/cjs/components/ModalTransition.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ModalTransitionConsumer = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var _React$createContext = /*#__PURE__*/_react.default.createContext({
  isOpen: true,
  onExited: function onExited() {}
}),
    Consumer = _React$createContext.Consumer,
    Provider = _React$createContext.Provider; // checks if children exist and are truthy


var hasChildren = function hasChildren(children) {
  return _react.default.Children.count(children) > 0 && _react.default.Children.map(children, function (child) {
    return !!child;
  }).filter(Boolean).length > 0;
};

var Transition = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(Transition, _React$Component);

  var _super = _createSuper(Transition);

  function Transition() {
    var _this;

    (0, _classCallCheck2.default)(this, Transition);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      currentChildren: undefined
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onExited", function () {
      _this.setState({
        currentChildren: _this.props.children
      });
    });
    return _this;
  }

  (0, _createClass2.default)(Transition, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement(Provider, {
        value: {
          onExited: this.onExited,
          isOpen: hasChildren(this.props.children)
        }
      }, this.state.currentChildren);
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      var previousChildren = state.currentChildren;
      var exiting = hasChildren(previousChildren) && !hasChildren(props.children);
      return {
        currentChildren: exiting ? previousChildren : props.children
      };
    }
  }]);
  return Transition;
}(_react.default.Component);

var ModalTransitionConsumer = Consumer;
exports.ModalTransitionConsumer = ModalTransitionConsumer;
var _default = Transition;
exports.default = _default;
