// @atlaskit/tag-group/dist/cjs/TagGroup/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = require("react");

var _core = require("@emotion/core");

/** @jsx jsx */
var baseStyles = (0, _core.css)({
  display: 'flex',
  width: '100%',
  flexWrap: 'wrap'
});
var justifyStartStyles = (0, _core.css)({
  justifyContent: 'flex-start'
});
var justifyEndStyles = (0, _core.css)({
  justifyContent: 'flex-end'
});
/**
 * __Tag group__
 *
 * A tag group controls the layout and alignment for a collection of tags.
 *
 * - [Examples](https://atlassian.design/components/tag-group/examples)
 * - [Code](https://atlassian.design/components/tag-group/code)
 * - [Usage](https://atlassian.design/components/tag-group/usage)
 */

var TagGroup = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var _ref$alignment = _ref.alignment,
      alignment = _ref$alignment === void 0 ? 'start' : _ref$alignment,
      children = _ref.children;
  return (0, _core.jsx)("div", {
    ref: ref,
    css: [baseStyles, alignment === 'start' && justifyStartStyles, alignment === 'end' && justifyEndStyles]
  }, children);
});
var _default = TagGroup;
exports.default = _default;
