// @atlaskit/checkbox/dist/esm/internal/styles.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n      & + svg {\n        --local-background: ", ";\n        --local-background-hover: ", ";\n        --local-background-active: ", ";\n        --local-background-checked: ", ";\n        --local-background-checked-hover: ", ";\n        --local-background-disabled: ", ";\n\n        --local-tick-rest: transparent;\n        --local-tick-checked: ", ";\n        --local-tick-disabled: ", ";\n        --local-tick-active: ", ";\n\n        --local-border: ", ";\n        --local-border-disabled: ", ";\n        --local-border-checked: ", ";\n        --local-border-active: ", ";\n        --local-border-invalid: ", ";\n        --local-border-focus: ", ";\n        --local-border-hover: ", ";\n        --local-border-checked-hover: ", ";\n        --local-border-checked-invalid: ", ";\n      }\n    "]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  /* Make the input invisible */\n  -webkit-appearance: none;\n  -moz-appearance: none;\n  margin: 0;\n\n  /*\n    Change the variables --checkbox-background-color, --checkbox-border-color\n    and --checkbox-tick-color according to user interactions.\n    All other variables are constant.\n    All styles from the input target the sibling svg.\n  */\n  & + svg {\n    --checkbox-background-color: var(--local-background);\n    --checkbox-border-color: var(--local-border);\n    --checkbox-tick-color: var(--local-tick-rest);\n    /* Color changes the background color */\n    color: var(--checkbox-background-color);\n    /* Fill changes the tick color */\n    fill: var(--checkbox-tick-color);\n    transition: color 0.2s ease-in-out, fill 0.2s ease-in-out;\n    flex-shrink: 0;\n    /* If the label is multiple lines, don't center the checkbox */\n    align-self: flex-start;\n  }\n  & + svg rect:first-of-type {\n    /* Stroke changes the color of the border */\n    stroke: var(--checkbox-border-color);\n    stroke-width: 2px;\n    transition: stroke 0.2s ease-in-out;\n  }\n\n  /* Styles are listed in order of increasing specificity  */\n\n  /*\n    Compiled will order the pseudo classes by LVFHA\n    As the background colour changes on hover, it is okay for the\n    border focus colour to take precedence over hover.\n    && is used to make the focus styles more specific than the hover\n    styles\n   */\n  &&:focus + svg,\n  &&:checked:focus + svg {\n    --checkbox-border-color: var(--local-border-focus);\n  }\n\n  &:hover + svg {\n    --checkbox-background-color: var(--local-background-hover);\n    --checkbox-border-color: var(--local-border-hover);\n  }\n  &:checked:hover + svg {\n    --checkbox-background-color: var(--local-background-checked-hover);\n    --checkbox-border-color: var(--local-border-checked-hover);\n  }\n\n  &:checked + svg {\n    --checkbox-background-color: var(--local-background-checked);\n    --checkbox-border-color: var(--local-border-checked);\n    --checkbox-tick-color: var(--local-tick-checked);\n  }\n\n  &[data-invalid] + svg {\n    --checkbox-border-color: var(--local-border-invalid);\n  }\n  &:checked[data-invalid] + svg {\n    --checkbox-border-color: var(--local-border-checked-invalid);\n  }\n\n  &:active + svg {\n    --checkbox-background-color: var(--local-background-active);\n    --checkbox-border-color: var(--local-border-active);\n  }\n  &:checked:active + svg {\n    --checkbox-tick-color: var(--local-tick-active);\n    --checkbox-background-color: var(--local-background-active);\n    --checkbox-border-color: var(--local-border-active);\n  }\n\n  &:disabled + svg,\n  &:disabled:hover + svg,\n  &:disabled:focus + svg,\n  &:disabled:active + svg,\n  &:disabled[data-invalid] + svg {\n    --checkbox-background-color: var(--local-background-disabled);\n    --checkbox-border-color: var(--local-border-disabled);\n    pointer-events: none;\n    cursor: not-allowed;\n  }\n\n  &:disabled:checked + svg {\n    --checkbox-tick-color: var(--local-tick-disabled);\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

import { css } from '@emotion/core';
import { B100, B200, B300, B400, B50, B75, DN10, DN200, DN30, DN80, DN90, N10, N20, N30, N40, N70, R300 } from '@atlaskit/theme/colors';
var checkboxStyles = css(_templateObject());
var checkboxThemeColors = {
  light: {
    borderColor: {
      rest: N40,
      disabled: N20,
      checked: B400,
      active: B50,
      invalid: R300,
      invalidAndChecked: R300,
      focused: B100,
      hovered: N40,
      hoveredAndChecked: B300
    },
    boxColor: {
      rest: N10,
      disabled: N20,
      active: B50,
      hoveredAndChecked: B300,
      hovered: N30,
      checked: B400
    },
    tickColor: {
      disabledAndChecked: N70,
      activeAndChecked: B400,
      checked: N10
    }
  },
  dark: {
    borderColor: {
      rest: DN80,
      disabled: DN10,
      checked: B400,
      active: B200,
      invalid: R300,
      invalidAndChecked: R300,
      focused: B75,
      hovered: DN200,
      hoveredAndChecked: B75
    },
    boxColor: {
      rest: DN10,
      disabled: DN10,
      active: B200,
      hoveredAndChecked: B75,
      hovered: DN30,
      checked: B400
    },
    tickColor: {
      disabledAndChecked: DN90,
      activeAndChecked: DN10,
      checked: DN10
    }
  }
};

var getCheckboxStyles = function getCheckboxStyles(mode) {
  var checkboxColors = checkboxThemeColors[mode];
  var boxColor = checkboxColors.boxColor;
  var tickColor = checkboxColors.tickColor;
  var borderColor = checkboxColors.borderColor;
  return [css(_templateObject2(), boxColor.rest, boxColor.hovered, boxColor.active, boxColor.checked, boxColor.hoveredAndChecked, boxColor.disabled, tickColor.checked, tickColor.disabledAndChecked, tickColor.activeAndChecked, borderColor.rest, borderColor.disabled, borderColor.checked, borderColor.active, borderColor.invalid, borderColor.focused, borderColor.hovered, borderColor.hoveredAndChecked, borderColor.invalidAndChecked), checkboxStyles];
};

export default getCheckboxStyles;
