// @compiled/utils/dist/source-maps.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildSourceMap = void 0;
var source_map_1 = require("source-map");
var convert_source_map_1 = __importDefault(require("convert-source-map"));
/**
 * Used to generate a inline source map for CSS.
 * It's input is the TypeScript source file,
 * an offset (where we should place the cursor when jumping to the source map)
 * and TypeScript context.
 *
 * Will return something like `/*# sourceMappingURL=...`
 */
function buildSourceMap(offset, code, fileName, sourceRoot) {
    var generator = new source_map_1.SourceMapGenerator({
        file: fileName,
        sourceRoot: sourceRoot,
    });
    generator.setSourceContent(fileName, code);
    generator.addMapping({
        generated: {
            line: 1,
            column: 0,
        },
        source: fileName,
        original: {
            line: offset.line + 1,
            column: offset.character,
        },
    });
    return convert_source_map_1.default.fromObject(generator).toComment({ multiline: true });
}
exports.buildSourceMap = buildSourceMap;
