// @compiled/react/dist/esm/codemods/styled-components-to-compiled/styled-components-to-compiled.js
import { hasImportDeclaration, convertDefaultImportToNamedImport, addCommentForUnresolvedImportSpecifiers, } from '../codemods-helpers';
var imports = {
    compiledStyledImportName: 'styled',
    styledComponentsPackageName: 'styled-components',
};
var transformer = function (fileInfo, _a, options) {
    var j = _a.jscodeshift;
    var source = fileInfo.source;
    var collection = j(source);
    var hasStyledComponentsImportDeclaration = hasImportDeclaration({
        j: j,
        collection: collection,
        importPath: imports.styledComponentsPackageName,
    });
    if (!hasStyledComponentsImportDeclaration) {
        return source;
    }
    addCommentForUnresolvedImportSpecifiers({
        j: j,
        collection: collection,
        importPath: imports.styledComponentsPackageName,
        allowedImportSpecifierNames: [],
    });
    convertDefaultImportToNamedImport({
        j: j,
        collection: collection,
        importPath: imports.styledComponentsPackageName,
        namedImport: imports.compiledStyledImportName,
    });
    return collection.toSource(options.printOptions || { quote: 'single' });
};
export default transformer;
