// @atlaskit/modal-dialog/dist/cjs/styled/Content.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionItem = exports.Actions = exports.Footer = exports.Body = exports.bodyStyles = exports.titleIconWrapperStyles = exports.TitleText = exports.Title = exports.Header = exports.wrapperStyles = exports.keylineHeight = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _core = require("@emotion/core");

var _styled = _interopRequireDefault(require("@emotion/styled"));

var _colors = require("@atlaskit/theme/colors");

var _components = require("@atlaskit/theme/components");

var _constants = require("@atlaskit/theme/constants");

var _math = require("@atlaskit/theme/math");

function _templateObject10() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  flex: 1 0 auto;\n  margin: 0 ", "px;\n"]);

  _templateObject10 = function _templateObject10() {
    return data;
  };

  return data;
}

function _templateObject9() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  display: inline-flex;\n  margin: 0 -", "px;\n"]);

  _templateObject9 = function _templateObject9() {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  position: relative;\n  align-items: center;\n  display: flex;\n  flex: 0 0 auto;\n  justify-content: space-between;\n  transition: box-shadow 200ms;\n  padding: ", "px ", "px ", "px\n    ", "px;\n  box-shadow: ", ";\n"]);

  _templateObject8 = function _templateObject8() {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  ", ";\n"]);

  _templateObject7 = function _templateObject7() {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  flex: 1 1 auto;\n  ", "\n  @media (min-width: 320px) and (max-width: 480px) {\n    overflow-y: auto;\n    height: 100%;\n  }\n  &:focus {\n    outline-offset: -1px;\n    outline-style: dotted;\n    outline-color: ", ";\n    outline-width: thin;\n  }\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  color: ", ";\n  margin-right: ", "px;\n  flex: 0 0 auto;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  flex: 1 1 auto;\n  min-width: 0;\n  word-wrap: break-word;\n  width: 100%;\n  ", ";\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  align-items: center;\n  display: flex;\n  font-size: 20px;\n  font-style: inherit;\n  font-weight: 500;\n  letter-spacing: -0.008em;\n  line-height: 1;\n  margin: 0;\n  min-width: 0;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  align-items: center;\n  display: flex;\n  flex: 0 0 auto;\n  justify-content: space-between;\n  transition: box-shadow 200ms;\n  position: relative;\n  padding: ", "px ", "px ", "px\n    ", "px;\n  box-shadow: ", ";\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  display: flex;\n  flex-direction: column;\n  flex: 1 1 auto;\n  max-height: 100%;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

// Constants
// ==============================
var modalPadding = (0, _constants.gridSize)() * 3;
var keylineColor = (0, _components.themed)({
  light: _colors.N30,
  dark: _colors.DN30
});
var keylineHeight = 2; // Wrapper
// ==============================

exports.keylineHeight = keylineHeight;
var wrapperStyles = (0, _core.css)(_templateObject()); // Header
// ==============================

exports.wrapperStyles = wrapperStyles;

var Header = _styled.default.header(_templateObject2(), modalPadding, modalPadding, modalPadding - keylineHeight, modalPadding, function (props) {
  return props.showKeyline ? "0 ".concat(keylineHeight, "px 0 0 ").concat(keylineColor(props)) : 'none';
});

exports.Header = Header;

var Title = _styled.default.h4(_templateObject3());

exports.Title = Title;

var TitleText = _styled.default.span(_templateObject4(), function (props) {
  return !props.isHeadingMultiline && "\n      white-space: nowrap;\n      overflow: hidden;\n      text-overflow: ellipsis;\n    ";
});

exports.TitleText = TitleText;
var iconColor = {
  danger: _colors.R400,
  warning: _colors.Y400
};

var titleIconWrapperStyles = function titleIconWrapperStyles(appearance) {
  return (0, _core.css)(_templateObject5(), iconColor[appearance], (0, _constants.gridSize)());
}; // Body
// ==============================

/**
  Adding the padding here avoids cropping box shadow on first/last
  children. The combined vertical spacing is maintained by subtracting the
  keyline height from header and footer.
*/


exports.titleIconWrapperStyles = titleIconWrapperStyles;

var bodyStyles = function bodyStyles(shouldScroll) {
  return (0, _core.css)(_templateObject6(), shouldScroll ? "\n        overflow-y: auto;\n        overflow-x: hidden;\n        padding: ".concat(keylineHeight, "px ").concat(modalPadding, "px;\n      ") : "\n        padding: 0 ".concat(modalPadding, "px;\n      "), _colors.N800);
};

exports.bodyStyles = bodyStyles;

// Body styles are used when customizing the body of the component.
// This is duplicated here: https://bitbucket.org/atlassian/atlassian-frontend/src/3c6a7cf3d18e97608fc1a4dc822b9bb003781bf2/packages/design-system/modal-dialog/src/components/Content.tsx#packages/design-system/modal-dialog/src/components/Content.tsx-341
// On the next major version delete `bodyStyles` from this declaration! They're just doubling up now!
var Body = _styled.default.div(_templateObject7(), function (props) {
  return bodyStyles(props.shouldScroll);
}); // Footer
// ==============================


exports.Body = Body;

var Footer = _styled.default.footer(_templateObject8(), modalPadding - keylineHeight, modalPadding, modalPadding, modalPadding, function (props) {
  return props.showKeyline ? "0 -".concat(keylineHeight, "px 0 0 ").concat(keylineColor(props)) : 'none';
});

exports.Footer = Footer;

var Actions = _styled.default.div(_templateObject9(), (0, _math.divide)(_constants.gridSize, 2));

exports.Actions = Actions;

var ActionItem = _styled.default.div(_templateObject10(), (0, _math.divide)(_constants.gridSize, 2));

exports.ActionItem = ActionItem;
