// @atlaskit/inline-dialog/dist/cjs/InlineDialog/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.InlineDialogWithoutAnalytics = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _reactNodeResolver = _interopRequireDefault(require("react-node-resolver"));

var _analyticsNext = require("@atlaskit/analytics-next");

var _popper = require("@atlaskit/popper");

var _version = require("../version.json");

var _styled = require("./styled");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var InlineDialog = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(InlineDialog, _Component);

  var _super = _createSuper(InlineDialog);

  function InlineDialog() {
    var _this;

    (0, _classCallCheck2.default)(this, InlineDialog);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClickOutside", function (event) {
      var _this$props = _this.props,
          isOpen = _this$props.isOpen,
          onClose = _this$props.onClose;

      if (event.defaultPrevented) {
        return;
      }

      var container = _this.containerRef;
      var trigger = _this.triggerRef;
      var target = event.target; // exit if we click outside but on the trigger — it can handle the clicks itself

      if (trigger && trigger.contains(target)) {
        return;
      } // call onClose if the click originated from outside the dialog


      if (isOpen && container && !container.contains(target)) {
        onClose && onClose({
          isOpen: false,
          event: event
        });
      }
    });
    return _this;
  }

  (0, _createClass2.default)(InlineDialog, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (typeof window === 'undefined') {
        return;
      }

      if (!prevProps.isOpen && this.props.isOpen) {
        window.addEventListener('click', this.handleClickOutside, true);
      } else if (prevProps.isOpen && !this.props.isOpen) {
        window.removeEventListener('click', this.handleClickOutside, true);
      }
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      if (typeof window === 'undefined') {
        return;
      }

      if (this.props.isOpen) {
        window.addEventListener('click', this.handleClickOutside, true);
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (typeof window === 'undefined') {
        return;
      }

      window.removeEventListener('click', this.handleClickOutside, true);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props2 = this.props,
          children = _this$props2.children,
          placement = _this$props2.placement,
          isOpen = _this$props2.isOpen,
          content = _this$props2.content,
          onContentBlur = _this$props2.onContentBlur,
          onContentFocus = _this$props2.onContentFocus,
          onContentClick = _this$props2.onContentClick,
          testId = _this$props2.testId;
      var popper = isOpen ? /*#__PURE__*/_react.default.createElement(_popper.Popper, {
        placement: placement
      }, function (_ref) {
        var ref = _ref.ref,
            style = _ref.style;
        return /*#__PURE__*/_react.default.createElement(_styled.Container, {
          onBlur: onContentBlur,
          onFocus: onContentFocus,
          onClick: onContentClick,
          innerRef: function innerRef(node) {
            _this2.containerRef = node;

            if (typeof ref === 'function') {
              ref(node);
            } else {
              ref.current = node;
            }
          },
          style: style,
          "data-testid": testId
        }, content);
      }) : null;
      return /*#__PURE__*/_react.default.createElement(_popper.Manager, null, /*#__PURE__*/_react.default.createElement(_popper.Reference, null, function (_ref2) {
        var ref = _ref2.ref;
        return /*#__PURE__*/_react.default.createElement(_reactNodeResolver.default, {
          innerRef: function innerRef(node) {
            _this2.triggerRef = node;

            if (typeof ref === 'function') {
              ref(node);
            } else {
              ref.current = node;
            }
          }
        }, children);
      }), popper);
    }
  }]);
  return InlineDialog;
}(_react.Component);

exports.InlineDialogWithoutAnalytics = InlineDialog;
(0, _defineProperty2.default)(InlineDialog, "defaultProps", {
  isOpen: false,
  onContentBlur: function onContentBlur() {},
  onContentClick: function onContentClick() {},
  onContentFocus: function onContentFocus() {},
  onClose: function onClose() {},
  placement: 'bottom-start'
});
var createAndFireEventOnAtlaskit = (0, _analyticsNext.createAndFireEvent)('atlaskit');

var _default = (0, _analyticsNext.withAnalyticsContext)({
  componentName: 'inlineDialog',
  packageName: _version.name,
  packageVersion: _version.version
})((0, _analyticsNext.withAnalyticsEvents)({
  onClose: createAndFireEventOnAtlaskit({
    action: 'closed',
    actionSubject: 'inlineDialog',
    attributes: {
      componentName: 'inlineDialog',
      packageName: _version.name,
      packageVersion: _version.version
    }
  })
})(InlineDialog));

exports.default = _default;
