// @atlaskit/modal-dialog/dist/esm/styled/Modal.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

function _templateObject7() {
  var data = _taggedTemplateLiteral(["\n  ", "\n  color: ", ";\n  display: flex;\n  flex-direction: column;\n  height: ", ";\n  max-height: 100%;\n  outline: 0;\n  pointer-events: auto;\n\n  @media (min-width: 320px) and (max-width: 480px) {\n    height: 100%;\n    max-height: 100%;\n    border-radius: 0;\n  }\n"]);

  _templateObject7 = function _templateObject7() {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = _taggedTemplateLiteral(["\n  ", ";\n  position: fixed;\n\n  @media (min-width: 320px) and (max-width: 480px) {\n    ", ";\n  }\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = _taggedTemplateLiteral(["\n  margin: ", "px auto;\n  position: relative;\n  width: ", ";\n  z-index: ", ";\n\n  @media (min-width: 320px) and (max-width: 480px) {\n    ", ";\n    margin: 0;\n  }\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteral(["\n  ", ";\n  position: absolute;\n\n  @media (min-width: 320px) and (max-width: 480px) {\n    ", ";\n  }\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteral(["\n  height: 100%;\n  left: 0;\n  position: fixed;\n  top: 0;\n  max-width: 100%;\n  width: 100%;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n  display: flex;\n  flex-direction: column;\n  height: ", ";\n  left: 0;\n  right: 0;\n  margin-left: auto;\n  margin-right: auto;\n  max-width: ", ";\n  top: ", "px;\n  width: ", ";\n  z-index: ", ";\n  pointer-events: none;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  height: 100vh;\n  left: 0;\n  overflow-y: auto;\n  position: absolute;\n  top: ", "px;\n  width: 100%;\n  z-index: ", ";\n  -webkit-overflow-scrolling: touch;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

import { css } from '@emotion/core';
import styled from '@emotion/styled';
import { DN50, N0, N30A, N60A, text } from '@atlaskit/theme/colors';
import { themed } from '@atlaskit/theme/components';
import { borderRadius, layers } from '@atlaskit/theme/constants';
import { gutter, WIDTH_ENUM } from '../shared-variables';

var boxShadow = function boxShadow(_ref) {
  var isChromeless = _ref.isChromeless;
  return isChromeless ? 'none' : "\n      0 0 0 1px ".concat(N30A, ", 0 2px 1px ").concat(N30A, ",\n      0 0 20px -6px ").concat(N60A, "\n    ");
};

var dialogBgColor = function dialogBgColor(_ref2) {
  var isChromeless = _ref2.isChromeless;
  return isChromeless ? 'transparent' : themed({
    light: N0,
    dark: DN50
  })();
};

var maxDimensions = "calc(100% - ".concat(gutter * 2, "px)");
var maxHeightDimensions = "calc(100% - ".concat(gutter * 2 - 1, "px)");
export var dialogWidth = function dialogWidth(_ref3) {
  var widthName = _ref3.widthName,
      widthValue = _ref3.widthValue;

  if (typeof widthValue === 'number') {
    return "".concat(widthValue, "px");
  }

  return widthName ? "".concat(WIDTH_ENUM.widths[widthName], "px") : widthValue || 'auto';
};
export var dialogHeight = function dialogHeight(_ref4) {
  var heightValue = _ref4.heightValue;

  if (typeof heightValue === 'number') {
    return "".concat(heightValue, "px");
  }

  return heightValue || 'auto';
};
/**
  NOTE:
  z-index
  - temporarily added to beat @atlaskit/navigation

  absolute + top
  - rather than fixed position so popper.js children are properly positioned

  overflow-y
  - only active when popper.js children invoked below the dialog
*/

export var FillScreen = styled.div(_templateObject(), function (props) {
  return props.scrollDistance;
}, layers.modal);

var positionBaseStyles = function positionBaseStyles(props) {
  return css(_templateObject2(), maxHeightDimensions, maxDimensions, gutter, dialogWidth(props), layers.modal());
};

var positionBaseResponsiveStyles = css(_templateObject3());
export var PositionerAbsolute = styled.div(_templateObject4(), positionBaseStyles, positionBaseResponsiveStyles);
export var PositionerRelative = styled.div(_templateObject5(), gutter, dialogWidth, layers.modal, positionBaseResponsiveStyles);
export var PositionerFixed = styled.div(_templateObject6(), positionBaseStyles, positionBaseResponsiveStyles);
export var Dialog = styled.div(_templateObject7(), function (props) {
  return props.isChromeless ? null : "\n          background-color: ".concat(dialogBgColor(props), ";\n          border-radius: ").concat(borderRadius(), "px;\n          box-shadow: ").concat(boxShadow(props), ";\n        ");
}, text, function (props) {
  return dialogHeight({
    heightValue: props.heightValue
  });
});
PositionerAbsolute.displayName = 'PositionerAbsolute';
Dialog.displayName = 'Dialog';
FillScreen.displayName = 'FillScreen';
PositionerRelative.displayName = 'PositionerRelative';
PositionerFixed.displayName = 'PositionerFixed';
