// @atlaskit/tag/dist/cjs/tag/internal/removable/remove-button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _core = require("@emotion/core");

var _close = _interopRequireDefault(require("@atlaskit/icon/glyph/editor/close"));

var _colors = require("@atlaskit/theme/colors");

var _constants = require("../../../constants");

/** @jsx jsx */
var baseStyles = (0, _core.css)({
  display: 'flex',
  height: '16px',
  margin: 0,
  padding: 0,
  position: 'absolute',
  right: 0,
  alignItems: 'center',
  justifyContent: 'center',
  alignSelf: 'center',
  appearance: 'none',
  backgroundColor: 'transparent',
  border: 'none',
  borderRadius: "var(".concat(_constants.cssVar.borderRadius, ")"),
  color: "var(--ds-text, ".concat(_colors.N500, ")"),
  cursor: 'pointer',
  pointerEvents: 'auto',
  '&::-moz-focus-inner': {
    margin: 0,
    padding: 0,
    border: 0
  },
  ':hover': {
    color: "var(".concat(_constants.cssVar.color.text.default, ")")
  }
});
var focusRingStyles = (0, _core.css)({
  '&:focus': {
    boxShadow: "0 0 0 2px var(".concat(_constants.cssVar.color.focusRing, ") inset"),
    outline: 'none'
  }
});

var RemoveButton = function RemoveButton(_ref) {
  var ariaLabel = _ref['aria-label'],
      onClick = _ref.onClick,
      onFocus = _ref.onFocus,
      onBlur = _ref.onBlur,
      onKeyPress = _ref.onKeyPress,
      onMouseOver = _ref.onMouseOver,
      onMouseOut = _ref.onMouseOut,
      testId = _ref.testId;
  return (0, _core.jsx)("button", {
    css: [baseStyles, focusRingStyles],
    tabIndex: 0,
    "aria-label": ariaLabel,
    onClick: onClick,
    onFocus: onFocus,
    onBlur: onBlur,
    onKeyPress: onKeyPress,
    onMouseOver: onMouseOver,
    onMouseOut: onMouseOut,
    type: "button",
    "data-testid": testId
  }, (0, _core.jsx)(_close.default, {
    label: "close tag",
    size: "small"
  }));
};

var _default = RemoveButton;
exports.default = _default;
