// @atlaskit/button/dist/esm/performance/examples.js
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import React, { useLayoutEffect, useRef, useState } from 'react';
import Button, { CustomThemeButton } from '../index'; // eslint-disable-next-line

import { fireEvent } from '@testing-library/react';

function Example(_ref) {
  var Component = _ref.Component;

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isSelected = _useState2[0],
      setIsSelected = _useState2[1];

  var _useState3 = useState(false),
      _useState4 = _slicedToArray(_useState3, 2),
      isDisabled = _useState4[0],
      setIsDisabled = _useState4[1];

  var ref = useRef(null);
  useLayoutEffect(function () {
    function toggleSelect() {
      setIsSelected(function (value) {
        return !value;
      });
    }

    function toggleDisabled() {
      setIsDisabled(function (value) {
        return !value;
      });
    }

    var el = ref.current;

    if (!el) {
      throw new Error('Could not find button ref');
    }

    el.addEventListener('toggle-select', toggleSelect);
    el.addEventListener('toggle-disabled', toggleDisabled);
    return function () {
      el.removeEventListener('toggle-select', toggleSelect);
      el.removeEventListener('toggle-disabled', toggleDisabled);
    };
  }, []);
  return (
    /*#__PURE__*/
    // @ts-ignore
    React.createElement(Component, {
      ref: ref,
      testId: "my-button",
      isSelected: isSelected,
      isDisabled: isDisabled,
      "data-is-selected": isSelected
    }, "Hello world")
  );
}

var interactionTasks = [{
  name: 'Click a button',
  description: 'Recording how long a mousedown + click event take to be processed',
  run: function () {
    var _run = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(_ref2) {
      var container, button;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              container = _ref2.container;
              button = container.querySelector('[data-testid="my-button"]');

              if (!(button == null)) {
                _context.next = 4;
                break;
              }

              throw new Error('Could not find button element');

            case 4:
              fireEvent.mouseDown(button);
              fireEvent.click(button);

            case 6:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    function run(_x) {
      return _run.apply(this, arguments);
    }

    return run;
  }()
}, {
  name: 'Focus on button',
  description: 'Focus on a button and wait for layout and paint to finish',
  run: function () {
    var _run2 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee3(_ref3) {
      var container, controls, button;
      return _regeneratorRuntime.wrap(function _callee3$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              container = _ref3.container, controls = _ref3.controls;
              button = container.querySelector('[data-testid="my-button"]');

              if (!(button == null)) {
                _context3.next = 4;
                break;
              }

              throw new Error('Could not find button element');

            case 4:
              _context3.next = 6;
              return controls.time( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
                return _regeneratorRuntime.wrap(function _callee2$(_context2) {
                  while (1) {
                    switch (_context2.prev = _context2.next) {
                      case 0:
                        fireEvent.focus(button);
                        fireEvent.blur(button);

                      case 2:
                      case "end":
                        return _context2.stop();
                    }
                  }
                }, _callee2);
              })));

            case 6:
            case "end":
              return _context3.stop();
          }
        }
      }, _callee3);
    }));

    function run(_x2) {
      return _run2.apply(this, arguments);
    }

    return run;
  }()
}, {
  name: 'Select button',
  description: 'Trigger the selection of a button',
  run: function () {
    var _run3 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee5(_ref5) {
      var container, controls, button;
      return _regeneratorRuntime.wrap(function _callee5$(_context5) {
        while (1) {
          switch (_context5.prev = _context5.next) {
            case 0:
              container = _ref5.container, controls = _ref5.controls;
              button = container.querySelector('[data-testid="my-button"]');

              if (!(button == null)) {
                _context5.next = 4;
                break;
              }

              throw new Error('Could not find button element');

            case 4:
              if (!(button.getAttribute('data-is-selected') !== 'false')) {
                _context5.next = 6;
                break;
              }

              throw new Error('Should not start selected');

            case 6:
              _context5.next = 8;
              return controls.time( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee4() {
                return _regeneratorRuntime.wrap(function _callee4$(_context4) {
                  while (1) {
                    switch (_context4.prev = _context4.next) {
                      case 0:
                        fireEvent(button, new Event('toggle-select'));

                      case 1:
                      case "end":
                        return _context4.stop();
                    }
                  }
                }, _callee4);
              })));

            case 8:
              if (!(button.getAttribute('data-is-selected') !== 'true')) {
                _context5.next = 10;
                break;
              }

              throw new Error('Should now be selected');

            case 10:
            case "end":
              return _context5.stop();
          }
        }
      }, _callee5);
    }));

    function run(_x3) {
      return _run3.apply(this, arguments);
    }

    return run;
  }()
}, {
  name: 'Disable button',
  description: 'Trigger the disabling of a button',
  run: function () {
    var _run4 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee7(_ref7) {
      var container, controls, button;
      return _regeneratorRuntime.wrap(function _callee7$(_context7) {
        while (1) {
          switch (_context7.prev = _context7.next) {
            case 0:
              container = _ref7.container, controls = _ref7.controls;
              button = container.querySelector('[data-testid="my-button"]');

              if (!(button == null)) {
                _context7.next = 4;
                break;
              }

              throw new Error('Could not find button element');

            case 4:
              if (!button.hasAttribute('disabled')) {
                _context7.next = 6;
                break;
              }

              throw new Error('Should not start disabled');

            case 6:
              _context7.next = 8;
              return controls.time( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee6() {
                return _regeneratorRuntime.wrap(function _callee6$(_context6) {
                  while (1) {
                    switch (_context6.prev = _context6.next) {
                      case 0:
                        fireEvent(button, new Event('toggle-disabled'));

                      case 1:
                      case "end":
                        return _context6.stop();
                    }
                  }
                }, _callee6);
              })));

            case 8:
              if (button.hasAttribute('disabled')) {
                _context7.next = 10;
                break;
              }

              throw new Error('Should now be disabled');

            case 10:
            case "end":
              return _context7.stop();
          }
        }
      }, _callee7);
    }));

    function run(_x4) {
      return _run4.apply(this, arguments);
    }

    return run;
  }()
}];
export var button = function button() {
  return /*#__PURE__*/React.createElement(Example, {
    Component: Button
  });
};
button.story = {
  name: 'Button',
  parameters: {
    performance: {
      interactions: interactionTasks
    }
  }
};
export var customThemeButton = function customThemeButton() {
  return /*#__PURE__*/React.createElement(Example, {
    Component: CustomThemeButton
  });
};
customThemeButton.story = {
  name: 'CustomThemeButton',
  parameters: {
    performance: {
      interactions: interactionTasks
    }
  }
};
