// @atlaskit/inline-message/dist/cjs/components/InlineMessage/styledInlineMessage.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Text = exports.Title = exports.ButtonContents = exports.Root = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var colors = _interopRequireWildcard(require("@atlaskit/theme/colors"));

var _components = require("@atlaskit/theme/components");

var _constants = require("../../constants");

function _templateObject5() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  color: ", ";\n  padding: 0 ", "px;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  white-space: nowrap;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  color: ", ";\n  font-weight: 500;\n  padding: 0 ", "px;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n      color: ", ";\n      text-decoration: underline;\n    "]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  align-items: center;\n  display: flex;\n  text-decoration: none;\n  ", ";\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  display: inline-block;\n  max-width: 100%;\n  &:focus {\n    outline: 1px solid ", ";\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var getFocusColor = (0, _components.themed)('appearance', {
  connectivity: {
    light: colors.B500,
    dark: colors.B200
  },
  confirmation: {
    light: colors.G400,
    dark: colors.G400
  },
  info: {
    light: colors.P500,
    dark: colors.P300
  },
  warning: {
    light: colors.Y500,
    dark: colors.Y500
  },
  error: {
    light: colors.R500,
    dark: colors.R500
  }
});

var Root = _styledComponents.default.div(_templateObject(), getFocusColor);

exports.Root = Root;

var ButtonContents = _styledComponents.default.div(_templateObject2(), function (_ref) {
  var isHovered = _ref.isHovered;
  return isHovered && (0, _styledComponents.css)(_templateObject3(), colors.N600);
});

exports.ButtonContents = ButtonContents;
var getTitleColor = (0, _components.themed)({
  light: colors.N600,
  dark: colors.DN600
});
var getTextColor = (0, _components.themed)({
  light: colors.N300,
  dark: colors.DN100
});

var Title = _styledComponents.default.span(_templateObject4(), getTitleColor, _constants.itemSpacing);

exports.Title = Title;

var Text = _styledComponents.default.span(_templateObject5(), getTextColor, _constants.itemSpacing);

exports.Text = Text;
