// @atlaskit/tabs/dist/cjs/components/Tabs.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TabsWithoutAnalytics = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = require("react");

var _core = require("@emotion/core");

var _analyticsNext = require("@atlaskit/analytics-next");

var _styles = require("../internal/styles");

var _version = require("../version.json");

var _TabContent = _interopRequireDefault(require("./TabContent"));

var _TabItem = _interopRequireDefault(require("./TabItem"));

var _TabsNavigation = _interopRequireDefault(require("./TabsNavigation"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var defaultIsSelectedTestNumber = function defaultIsSelectedTestNumber(selectedIndex, _tab, tabIndex) {
  return selectedIndex === tabIndex;
};

var defaultIsSelectedTestObject = function defaultIsSelectedTestObject(selected, tab) {
  return selected === tab;
};

var defaultComponents = {
  Content: _TabContent.default,
  Item: _TabItem.default
};

var Tabs = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Tabs, _Component);

  var _super = _createSuper(Tabs);

  function Tabs(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Tabs);
    _this = _super.call(this, props);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "resolveSelected", function (selected, newProps) {
      var _ref = newProps || _this.props,
          tabs = _ref.tabs,
          isSelectedTest = _ref.isSelectedTest;

      var testFunction = function () {
        if (isSelectedTest) {
          return isSelectedTest;
        }

        if (typeof selected === 'number') {
          return defaultIsSelectedTestNumber;
        }

        return defaultIsSelectedTestObject;
      }();

      return tabs.find(function (tab, tabIndex) {
        return testFunction(selected, tab, tabIndex);
      }) || tabs[0];
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onSelect", function (newselected, newSelectedIndex) {
      var _this$props = _this.props,
          onSelect = _this$props.onSelect,
          selected = _this$props.selected;

      if (typeof onSelect === 'function') {
        onSelect(newselected, newSelectedIndex);
      }

      if (typeof selected === 'undefined') {
        _this.setState({
          selected: newselected
        });
      }
    });
    var initiallyselected = _this.props.selected || _this.props.defaultSelected || _this.props.tabs[0];

    var _selected = _this.resolveSelected(initiallyselected);

    _this.state = {
      selected: _selected
    };
    return _this;
  }

  (0, _createClass2.default)(Tabs, [{
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(newProps) {
      if (typeof newProps.selected !== 'undefined' && newProps.selected !== this.state.selected) {
        var selected = this.resolveSelected(newProps.selected, newProps);
        this.setState({
          selected: selected
        });
      } else if (newProps.tabs !== this.props.tabs) {
        var updatedselected = this.resolveSelected(this.state.selected, newProps);
        this.setState({
          selected: updatedselected
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          components = _this$props2.components,
          tabs = _this$props2.tabs,
          testId = _this$props2.testId,
          isContentPersisted = _this$props2.isContentPersisted;
      var selected = this.state.selected;

      var _defaultComponents$co = _objectSpread(_objectSpread({}, defaultComponents), components),
          Content = _defaultComponents$co.Content,
          Item = _defaultComponents$co.Item;

      var contents = function contents() {
        return tabs.map(function (tab, index) {
          var isSelected = tab === selected;
          var contentProps = {
            data: tab,
            elementProps: {
              role: 'tabpanel'
            },
            isSelected: isSelected
          };
          return (0, _core.jsx)(Content, (0, _extends2.default)({
            key: index
          }, contentProps));
        });
      };

      var contentProps = {
        data: selected,
        elementProps: {
          role: 'tabpanel'
        },
        isSelected: true
      };
      return (0, _core.jsx)("div", {
        "data-testid": testId,
        css: _styles.tabStyles
      }, (0, _core.jsx)(_TabsNavigation.default, {
        component: Item,
        onSelect: this.onSelect,
        selected: selected,
        tabs: tabs
      }), isContentPersisted ? contents() : (0, _core.jsx)(Content, contentProps));
    }
  }]);
  return Tabs;
}(_react.Component);

exports.TabsWithoutAnalytics = Tabs;
(0, _defineProperty2.default)(Tabs, "defaultProps", {
  components: {},
  isContentPersisted: false
});
var createAndFireEventOnAtlaskit = (0, _analyticsNext.createAndFireEvent)('atlaskit');

var _default = (0, _analyticsNext.withAnalyticsContext)({
  componentName: 'tabs',
  packageName: _version.name,
  packageVersion: _version.version
})((0, _analyticsNext.withAnalyticsEvents)({
  onSelect: createAndFireEventOnAtlaskit({
    action: 'clicked',
    actionSubject: 'tab',
    attributes: {
      componentName: 'tabs',
      packageName: _version.name,
      packageVersion: _version.version
    }
  })
})(Tabs));

exports.default = _default;
