// @atlaskit/tooltip/dist/cjs/TooltipPrimitive.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = require("react");

var _core = require("@emotion/core");

var _constants = require("@atlaskit/theme/constants");

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2.default)(["\n  z-index: ", ";\n  pointer-events: none;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var primitiveCss = (0, _core.css)(_templateObject(), _constants.layers.tooltip());
var TooltipPrimitive = /*#__PURE__*/(0, _react.forwardRef)(function TooltipPrimitive(_ref, ref) {
  var style = _ref.style,
      className = _ref.className,
      children = _ref.children,
      placement = _ref.placement,
      testId = _ref.testId;
  return (0, _core.jsx)("div", {
    role: "tooltip",
    ref: ref,
    style: style,
    className: className,
    css: primitiveCss,
    "data-placement": placement,
    "data-testid": testId
  }, children);
});
TooltipPrimitive.displayName = 'TooltipPrimitive';
var _default = TooltipPrimitive;
exports.default = _default;
