// @atlaskit/tag/dist/cjs/tag/internal/simple/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = require("react");

var _core = require("@emotion/core");

var _base = _interopRequireDefault(require("../shared/base"));

var _before = _interopRequireDefault(require("../shared/before"));

var _content = _interopRequireDefault(require("../shared/content"));

/** @jsx jsx */
var SimpleTag = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var appearance = _ref.appearance,
      _ref$elemBefore = _ref.elemBefore,
      elemBefore = _ref$elemBefore === void 0 ? null : _ref$elemBefore,
      _ref$color = _ref.color,
      color = _ref$color === void 0 ? 'standard' : _ref$color,
      href = _ref.href,
      testId = _ref.testId,
      _ref$text = _ref.text,
      text = _ref$text === void 0 ? '' : _ref$text,
      linkComponent = _ref.linkComponent;
  var content = (0, _core.jsx)(_content.default, {
    elemBefore: elemBefore,
    isRemovable: false,
    text: text,
    color: color,
    href: href,
    linkComponent: linkComponent
  });
  return (0, _core.jsx)(_base.default, {
    ref: ref,
    testId: testId,
    href: href,
    appearance: appearance,
    color: color,
    before: (0, _core.jsx)(_before.default, {
      elemBefore: elemBefore
    }),
    contentElement: content
  });
});

var _default = /*#__PURE__*/(0, _react.memo)(SimpleTag);

exports.default = _default;
