// @atlaskit/modal-dialog/dist/esm/components/FocusLock.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React from 'react';
import ReactFocusLock from 'react-focus-lock';
import invariant from 'tiny-invariant';

// Thin wrapper over react-focus-lock. This wrapper only exists to ensure API compatibility.
// This component should be deleted during https://ecosystem.atlassian.net/browse/AK-5658
var FocusLock = /*#__PURE__*/function (_React$Component) {
  _inherits(FocusLock, _React$Component);

  var _super = _createSuper(FocusLock);

  function FocusLock() {
    _classCallCheck(this, FocusLock);

    return _super.apply(this, arguments);
  }

  _createClass(FocusLock, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this$props = this.props,
          isEnabled = _this$props.isEnabled,
          autoFocus = _this$props.autoFocus;

      if (process.env.NODE_ENV !== 'production') {
        invariant(typeof autoFocus === 'boolean', '@atlaskit/modal-dialog: Passing a function as autoFocus is deprecated. Instead call focus on the element ref or use the autofocus property.');
      }

      if (typeof autoFocus === 'function' && isEnabled) {
        var elem = autoFocus();

        if (elem && elem.focus) {
          elem.focus();
        }
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          isEnabled = _this$props2.isEnabled,
          autoFocus = _this$props2.autoFocus,
          shouldReturnFocus = _this$props2.shouldReturnFocus;
      return /*#__PURE__*/React.createElement(ReactFocusLock, {
        disabled: !isEnabled,
        autoFocus: !!autoFocus,
        returnFocus: shouldReturnFocus
      }, this.props.children);
    }
  }]);

  return FocusLock;
}(React.Component);

_defineProperty(FocusLock, "defaultProps", {
  autoFocus: true,
  isEnabled: true,
  shouldReturnFocus: true
});

export { FocusLock as default };
